File: src/csharp/examples/cublas/Axpy.cs
using Alea.CUDA;
using Alea.CUDA.CULib;
using NUnit.Framework;

namespace Tutorial.Cs.examples.cublas
{
    class Axpy
    {
        private static readonly CUBLAS Cublas = CUBLAS.Default;
        private static readonly Worker Worker = Cublas.Worker;

        //[axpyCpu]
        public static class cpu
        {
            public static double[] Daxpy(int n, double alpha, double[] x, int incx, double[] y, int incy)
            {
                var r = new double[n];
                for (var i = 0; i < n; ++i)
                {
                    r[i] = alpha * x[i] + y[i];
                }
                return r;
            }

            public static double2[] Zaxpy(int n, double2 alpha, double2[] _x, int incx, double2[] _y, int incy)
            {
                var x = double2.ToComplex(_x);
                var y = double2.ToComplex(_y);
                var r = double2.ToComplex(new double2[n]);
                for (var i = 0; i < n; ++i)
                {
                    r[i] = alpha.ToComplex() * x[i] + y[i];
                }
                return double2.OfComplex(r);
            }
        }
        //[/axpyCpu]

        //[axpyGpu]
        public static class gpu
        {
            public static double[] Daxpy(int n, double alpha, double[] x, int incx, double[] y, int incy)
            {
                using (var dalpha = Worker.Malloc(new[] { alpha }))
                using (var dx = Worker.Malloc(x))
                using (var dy = Worker.Malloc(y))
                {
                    Cublas.Daxpy(n, dalpha.Ptr, dx.Ptr, incx, dy.Ptr, incy);
                    return dy.Gather();
                }
            }

            public static double2[] Zaxpy(int n, double2 alpha, double2[] x, int incx, double2[] y, int incy)
            {
                using (var dalpha = Worker.Malloc(new[] { alpha }))
                using (var dx = Worker.Malloc(x))
                using (var dy = Worker.Malloc(y))
                {
                    Cublas.Zaxpy(n, dalpha.Ptr, dx.Ptr, incx, dy.Ptr, incy);
                    return dy.Gather();
                }
            }
        }
        //[/axpyGpu]
    }   
    
    //[axpyTest]
    public partial class Test
    {
        private const int Incx = 1;
        private const int Incy = 1;

        [Test]
        public static void DaxpyTest()
        {
            Util.FallBack(() =>
            {
                const int N = 5;
                const double alpha = 2.0;

                var x = new double[N];
                var y = new double[N];

                for (var i = 0; i < N; ++i)
                {
                    x[i] = 2.0;
                    y[i] = 1.0;
                }

                var outputs = Axpy.gpu.Daxpy(N, alpha, x, Incx, y, Incy);
                var expected = Axpy.cpu.Daxpy(N, alpha, x, Incx, y, Incy);

                for (var i = 0; i < N; ++i)
                    Assert.AreEqual(outputs[i], expected[i], 1e-12);
            });
        }

        [Test]
        public static void ZaxpyTest()
        {
            Util.FallBack(() =>
            {
                const int N = 5;
                var alpha = new double2(2.0, 2.0);

                var x = new double2[N];
                var y = new double2[N];

                var outputs = Axpy.gpu.Zaxpy(N, alpha, x, Incx, y, Incy);
                var expected = Axpy.cpu.Zaxpy(N, alpha, x, Incx, y, Incy);

                for (var i = 0; i < N; ++i)
                {
                    Assert.AreEqual(outputs[i].x, expected[i].x, 1e-12);
                    Assert.AreEqual(outputs[i].y, expected[i].y, 1e-12);
                }
            });
        }
    }
    //[/axpyTest]
}

File: src/fsharp/examples/cublas/Axpy.fs
(*** hide ***)
module Tutorial.Fs.examples.cublas.Axpy

open System
open Alea.CUDA
open Alea.CUDA.Utilities
open Alea.CUDA.CULib
open NUnit.Framework
open FsUnit

(**
# Axpy

Here we see a basic example of how to use the cuBLAS library.
The function [<t>axpy](http://docs.nvidia.com/cuda/cublas/#cublas-lt-t-gt-axpy) multiplies the vector 
x by the scalar alpha and adds it to the vector y.
*)

(**
To use the cuBLAS library from Alea GPU you simply need to open `Alea.CUDA.cuBLAS`.
The default implementation is available in `Alea.CUDA.CULib.CUBLAS.Default`.
*)
let cublas = CULib.CUBLAS.Default

(**
First, we implement a simple CPU version of the function `<t>axpy`.
*)
(*** define:axpyCpu ***)
module cpu = 
    let daxpy n alpha (x:float[]) incx (y:float[]) incy =
        [for i in [0..n-1] -> alpha * x.[i] + y.[i]]
        |> Array.ofList

    let zaxpy n (alpha:double2) (x:double2[]) incx (y:double2[]) incy =
        let alpha = alpha.ToComplex()
        let x = double2.ToComplex(x)
        let y = double2.ToComplex(y)
        [for i in [0..n-1] -> alpha * x.[i] + y.[i]]
        |> Array.ofList

(**
The GPU implementation of `<t>axpy` is quite simple. Because cuBlas functions are executed from the host side, 
all you need to do is to allocate the device memory with a worker.
*)
(*** define:axpyGpu ***)
module gpu =
    let daxpy n (alpha:float) (x:float[]) incx (y:float[]) incy =
        let worker = Worker.Default
        use dalpha = worker.Malloc([|alpha|])
        use dx = worker.Malloc(x)
        use dy = worker.Malloc(y)
        cublas.Daxpy(n, dalpha.Ptr, dx.Ptr, incx, dy.Ptr, incy)
        dy.Gather()

    let zaxpy n (alpha:double2) (x:double2[]) incx (y:double2[]) incy =
        let worker = Worker.Default
        use dalpha = worker.Malloc([|alpha|])
        use dx = worker.Malloc(x)
        use dy = worker.Malloc(y)
        cublas.Zaxpy(n, dalpha.Ptr, dx.Ptr, incx, dy.Ptr, incy)
        dy.Gather()

(**
When using cuBLAS in your own implementations, review the official cuBLAS 
documentation and make sure you are gathering the appropriate data from the GPU.

Testing against CPU reference implementation.
*)
(*** define:axpyTest ***)
[<Test>]
let daxpyTest() =
    Util.fallback <| fun _ ->
        let n = 5
        let incx,incy = 1,1
        let alpha = 2.0
        let x = Array.create n 2.0
        let y = Array.create n 1.0
    
        let outputs = gpu.daxpy n alpha x incx y incy
        let expected = cpu.daxpy n alpha x incx y incy

        printfn "cpu result: %A" expected
        printfn "gpu result: %A" outputs

        outputs |> should equal expected

[<Test>]
let zaxpyTest() =
    Util.fallback <| fun _ ->
        let n = 5
        let incx,incy = 1,1
        let alpha = double2(2.0, 2.0)
        let x = Array.init n (fun _ -> double2(2.0, 0.5))
        let y = Array.init n (fun _ -> double2(1.0, 0.5))
    
        let outputs = gpu.zaxpy n alpha x incx y incy |> Array.map (fun x -> Numerics.Complex(x.x, x.y))
        let expected = cpu.zaxpy n alpha x incx y incy

        printfn "cpu result: %A" expected
        printfn "gpu result: %A" outputs
    
        outputs |> should equal expected
